coins: remove logic for spent-and-FRESH cache entries and writing non-DIRTY entries

andrewtoth opened the pull request:
Following up from https://github.com/bitcoin/bitcoin/pull/28280#discussion_r1655477509, which suggested a revival of #18746.

`GetCoin` will never return `true` for a spent entry, so we can safely assume that any entry we fetch will not be spent. This lets us remove the only non-test code path which adds a FRESH-but-not-DIRTY entry to the flagged linked list. This in turn ensures all entries being sent to `BatchWrite` are `DIRTY` entries.

A corollary is that all spent coins must be DIRTY. The only time a coin can be spent and not DIRTY is when the `CCoinsViewCacheEntry` is created with an empty coin. The last commit makes this more clear by checking for freshness if inserted instead of if spent and not DIRTY.

This is a pure refactor removing dead code which handles a non-existent corner case of a FRESH-but-not-DIRTY entry in the `CCoinsViewCache` or a spent entry in `CCoinsViewDB`. There is a lot of test code which tries to exercise this corner case, which is also updated in this PR to behave like non-test code.

andrewtoth committed:
coins: coins returned from GetCoin cannot be spent

andrewtoth committed:
coins: throw errors for non-DIRTY or spent and FRESH coins in BatchWrite

There are no code paths which add a non-DIRTY entry to the cursor in
BatchWrite, so we can safely make this a logic error and test for it.

There are no code paths which add a spent and FRESH coin to the cursor
in BatchWrite, so we can safely make this a logic error and test for it.

andrewtoth committed:
coins: remove redundant IsDirty checks in BatchWrite

It is no longer possible to get non-DIRTY entries in BatchWrite

andrewtoth committed:
coins: remove IsFresh check in Uncache

It is not possible for an entry to be FRESH if not already DIRTY.

paplorinc commented on src/test/fuzz/coinscache_sim.cpp:
```diff
@@ -151,14 +148,10 @@ class CoinsViewBottom final : public CCoinsView
     {
```
I was surprise to see that the method called `GetCoin` now:
* overwrites the dummy coin parameter with the value found in the cache
* returns a boolean indicating whether the coin exists in the cache and is not spent (the name doesn't have any hints about it being spent or not)

I don't mind if you think it's outside the scope of the current change, but checking the usages it seems to me we can now simplify `GetCoin` to actually return the coin and let the callsite check for spent and it might simplify the usages as well:
```C++
std::optional<Coin> CCoinsViewCache::GetCoin(const COutPoint& outpoint) const
{
    auto it = FetchCoin(outpoint);
    if (it != cacheCoins.end()) {
        return it->second.coin;
    }
    return std::nullopt;
}
```

<details>
<summary>... and in tests</summary>
```C++
std::optional<Coin> GetCoin(const COutPoint& outpoint) const
{
    if (const auto [it, found] = m_data.find(outpoint); found) {
        return it->second;
    }
    return std::nullopt;
}
```
</details>

Which would simplify usages from e.g.
```C++
Coin coin_using_get_coin;
const bool exists_using_get_coin = coins_view_cache.GetCoin(random_out_point, coin_using_get_coin);
if (exists_using_get_coin) {
    assert(coin_using_get_coin == coin_using_access_coin);
}
assert((exists_using_access_coin && exists_using_have_coin_in_cache && exists_using_have_coin && exists_using_get_coin) ||
       (!exists_using_access_coin && !exists_using_have_coin_in_cache && !exists_using_have_coin && !exists_using_get_coin));
```
to something like:
```C++
if (auto coin_optional = coins_view_cache.GetCoin(random_out_point); coin_optional && !coin_optional->IsSpent()) {
    assert(*coin_optional == coin_using_access_coin);
    assert(exists_using_access_coin && exists_using_have_coin_in_cache && exists_using_have_coin);
} else {
    assert(!exists_using_access_coin && !exists_using_have_coin_in_cache && !exists_using_have_coin);
}
```

What do you think?

andrewtoth replied:
This behavior is documented [here](https://github.com/bitcoin/bitcoin/blob/master/src/coins.h#L306-L310), so not sure why test code bothered to do this. Indeed, switching to `std::optional` would be cleaner, but I think a much bigger refactor. This is also suggested in the original PR [here](https://github.com/bitcoin/bitcoin/pull/18746#issuecomment-842393167).

paplorinc replied:
> but I think a much bigger refactor

Maybe, would it help if I did that in a different PR?

andrewtoth replied:
Yes, but it would conflict heavily with this one. Many of the tests would have to be removed as well.

paplorinc replied:
I'm fine with that, extracted to a branch for now, will push as a draft after your PR stabilizes

paplorinc replied:
you can resolve this comment, I see now that this should definitely be investigated in a separate PR (it's a big change).

andrewtoth replied:
I think you can resolve your own comment threads too 😉

paplorinc replied:
I don't have write access to yiur PR:
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/about-pull-request-reviews#resolving-conversations

paplorinc commented on src/coins.h:
```diff
@@ -184,13 +177,13 @@ struct CCoinsCacheEntry
-    //! Only call Next when this entry is DIRTY, FRESH, or both
+    //! Only call Next when this entry is DIRTY
     inline CoinsCachePair* Next() const noexcept {
         Assume(m_flags);
```
```suggestion
        Assume(IsDirty());
```

nit: it's a cheap call, why not make it an assert instead?

andrewtoth replied:
Done.

paplorinc commented on src/test/coins_tests.cpp:
```diff
@@ -53,14 +53,11 @@ class CCoinsViewTest : public CCoinsView
     bool BatchWrite(CoinsViewCacheCursor& cursor, const uint256& hashBlock) override
     {
-        for (auto it{cursor.Begin()}; it != cursor.End(); it = cursor.NextAndMaybeErase(*it)){
-            if (it->second.IsDirty()) {
```
is this is always the case, consider adding an assertion
```C++
assert(it->second.IsDirty());
```

(we can remove them in the last commit, but would be useful for the intermediary commits at least)

andrewtoth replied:
Done.

paplorinc commented on src/coins.cpp:
```diff
@@ -275,7 +275,7 @@ bool CCoinsViewCache::Sync()
 void CCoinsViewCache::Uncache(const COutPoint& hash)
 {
     CCoinsMap::iterator it = cacheCoins.find(hash);
-    if (it != cacheCoins.end() && !it->second.IsDirty() && !it->second.IsFresh()) {
```
same, can we add a (temporary) assertion for the removed value?

l0rinc reviewed (commented):
The code became a lot simpler now - have a few suggestion I'd like us to consider, but I understand if you think it's out of scope.

Once we get to a state that is ACK-worthy, I'll enable hard assertions everywhere and run an IBD and check if it crashes anywhere.

paplorinc commented on src/coins.cpp:
```diff
@@ -49,10 +49,8 @@ CCoinsMap::iterator CCoinsViewCache::FetchCoin(const COutPoint &outpoint) const
-            // The parent only has an empty entry for this outpoint; we can consider our version as fresh.
-            ret->second.AddFlags(CCoinsCacheEntry::FRESH, *ret, m_sentinel);
-        }
+        // GetCoin will never return true if the coin is spent
```
This doesn't sound like it's `GetCoin`'s responsibility, I think the call-sites should be checking for whether the coins are spent or not, not the cache accessor method

andrewtoth replied:
It's specifically documented that this is the behavior. Why shouldn't we be able to rely on it?

paplorinc replied:
If we need to document the behavior of a method, it suggests that we’ve acknowledged it may not be intuitive for everyone, indicating that we’ve conceded on finding a better design.
I think there's a way we can avoid the comment: by separating the two concerns of getting a value from cache and checking if it's spent or not.
Or if you insist, we can call it "GetUnspent" in which case we can also get rid of the comment and call-site asserts and keep the current behavior.
What do you think?

andrewtoth replied:
What you are suggesting is basically a concept NACK on this PR. If we aren't confident that we will never get a spent coin from GetCoin, then we need to keep the logic to handle spent and FRESH coins.

paplorinc replied:
I won't block this since I'm fine with discussing/addressing any issues in later commits as well. 
I also don't yet understand the second order effects of these changes, so treat my comments as curious observations.

So my concern is that in this PR was that now we're tying `coin.IsSpent()` with retrieving values from the cache.
So if my understanding is correct we should either:
* be able to get a coin from the cache and ignore/assert whether it's spent or not - e.g. [`HaveCoin`](https://github.com/bitcoin/bitcoin/blob/master/src/coins.cpp#L21) will now return false when the cache contains it but it's spent... and if that's not possible in certain situations I would assert instead of returning the boolean.
So in that case `GetCoin` would just get it from the cache and the call-site will decide whether it should assert for `IsSpent` or branch based on it, something like this: https://github.com/bitcoin/bitcoin/blob/9db8b30788eff2b8e9c9f629a6a03886a28a3281/src/test/fuzz/coins_view.cpp#L167
* rename the method to reflect the new responsibility that it has

There's also a third possibility that I'm completely off on what this PR does - your patience is appreciated :)

andrewtoth replied:
@l0rinc It seems to me like you are thinking this PR has any behavior change. Any coin retrieved from `GetCoin` will be unspent if `GetCoin` returns true. You can follow this logically from seeing the return value [here](https://github.com/bitcoin/bitcoin/blob/master/src/coins.cpp#L65) and then the only base class is `CCoinsViewDB` which does not save spent coins so cannot retrieve them in `GetCoin`.

This PR is recognizing that `FetchCoin` logically cannot get to [here](https://github.com/bitcoin/bitcoin/blob/master/src/coins.cpp#L52) with a spent coin. [This](https://github.com/bitcoin/bitcoin/blob/master/src/coins.cpp#L54) is a dead code path.

l0rinc replied:
Yeah, I understand that.

What I'm struggling with is the cache retrieval, and especially the invalidation logic, which seems to be scattered throughout the code and tightly coupled with the business logic.

Would you be open to a screen-sharing session to discuss this? It would help me better understand the constraints and likely speed up my review process as well.

l0rinc replied:
So my understanding that if a coin is already spent it's basically 
"less likely" to be in the cache (0% without reorgs, right?), so we want to get rid of it ASAP.
Given that the complete UTXO set doesn't fit into memory anymore, would it maybe make sense to have a more complex retainment heuristic for each cache layer, e.g. recency, dust amount, unspendables, time-locked, etc?

andrewtoth replied:
> Given that the complete UTXO set doesn't fit into memory anymore, would it maybe make sense to have a more complex retainment heuristic for each cache layer,

Since we now have a linked list of entries, we could have another sentinel that tracks coins that have their flags cleared. Then, when doing a flush we could just remove a chunk from the bottom of the linked list (entries added least recently).

l0rinc replied:
Added https://github.com/bitcoin/bitcoin/pull/30849 to somewhat mitigate the `GetCoin` mutations in tests - hoping this will get us closer to excluding invalid tests states.

andrewtoth commented:
> Once we get to a state that is ACK-worthy, I'll enable hard assertions everywhere and run an IBD and check if it crashes anywhere.

I think for a change like this, spending time fuzzing would be more valuable. I believe it already runs in debug mode for that so `Assume`s would trigger crashes.

l0rinc commented on src/coins.cpp:
```diff
@@ -184,34 +182,33 @@ void CCoinsViewCache::SetBestBlock(const uint256 &hashBlockIn) {
-        // Ignore non-dirty entries (optimization).
         if (!it->second.IsDirty()) {
-            continue;
+            throw std::logic_error("A non-DIRTY coin was returned from the cursor in BatchWrite");
+        }
+        if (it->second.IsFresh() && it->second.coin.IsSpent()) {
+            throw std::logic_error("A FRESH coin was not removed when it was spent");
         }
```
I find it inconsistent that we're throwing here only for testability - since we assume this cannot happen in reality.
Basically, this is another "return value" of the method, so declaring it a `bool` was essentially a lie (especially since it always seems to just return `true`).
Could we use the return value instead to signal failure?

andrewtoth replied:
This can happen in our fuzz tests. I think a follow up to this change would be to remove direct access to the flags, and replace with methods `SetDirty` and `SetDirtyAndFresh`. Then we can get rid of fuzzing all kinds of flags and simplify this.

l0rinc replied:
Can we consider fixing that before this PR? I wouldn't mind doing parts of that myself!

andrewtoth replied:
Are you suggesting to refactor direct flags access with `SetDirty` and `SetDirtyAndFresh` first? If we do that first, we would still need a `SetFresh` setter since we still have a case here with setting FRESH-but-not-DIRTY. It's kind of a chicken-and-egg problem.

l0rinc replied:
I suggest we make our tests reflect reality before we do this change, since currently if we add asserts about the internal consistency of the code, the tests seem to constantly violate them.

l0rinc commented on src/coins.cpp:
```diff
@@ -184,34 +182,33 @@ void CCoinsViewCache::SetBestBlock(const uint256 &hashBlockIn) {
+                entry.coin = it->second.coin;
+            }
+            cachedCoinsUsage += entry.coin.DynamicMemoryUsage();
+            entry.AddFlags(CCoinsCacheEntry::DIRTY, *itUs, m_sentinel);
+            // We can mark it FRESH in the parent if it was FRESH in the child
+            // Otherwise it might have just been flushed from the parent's cache
+            // and already exist in the grandparent
+            if (it->second.IsFresh()) {
+                entry.AddFlags(CCoinsCacheEntry::FRESH, *itUs, m_sentinel);
             }
```
Instead of calling `AddFlags` twice, in [other cases](https://github.com/bitcoin/bitcoin/blob/master/src/coins.cpp#L100) we've compressed this into a single line:
```suggestion
            // We can mark it FRESH in the parent if it was FRESH in the child
            // Otherwise it might have just been flushed from the parent's cache
            // and already exist in the grandparent
            entry.AddFlags(CCoinsCacheEntry::DIRTY | (it->second.IsFresh() ? CCoinsCacheEntry::FRESH : 0), *itUs, m_sentinel);
```

andrewtoth replied:
I think I only touch this in whitespace, so I don't want to refactor this right now.

l0rinc commented on src/coins.cpp:
```diff
@@ -184,34 +182,33 @@ void CCoinsViewCache::SetBestBlock(const uint256 &hashBlockIn) {
-                // We can mark it FRESH in the parent if it was FRESH in the child
-                // Otherwise it might have just been flushed from the parent's cache
-                // and already exist in the grandparent
-                if (it->second.IsFresh()) {
-                    entry.AddFlags(CCoinsCacheEntry::FRESH, *itUs, m_sentinel);
-                }
+            // Create the coin in the parent cache, move the data up
+            // and mark it as dirty.
+            itUs = cacheCoins.try_emplace(it->first).first;
```
Now that this structure is simpler, we could modernize (and optimize) this `find` / `try_emplace` construct as:
```C++
auto [itUs, inserted] = cacheCoins.try_emplace(it->first);
if (inserted) {
```

andrewtoth replied:
Yes, that would be a good follow up!

l0rinc reviewed (changes requested):
Dug deeper and added more relevant questions - bear with me if they're outside the scope of this PR.

andrewtoth committed:
coins: assume entry is dirty in Next and Prev

andrewtoth committed:
test: simplify coins sanity check

andrewtoth committed:
coins: set all inserted spent coins to fresh

A spent coins must be DIRTY, so remove
references to spent but not DIRTY coins.
The only way a spent coin can be not DIRTY
is when creating the CCoinsCacheEntry with
an empty coin. This can be made more clear
by setting fresh if inserted, instead of
checking if an unspent coin is not DIRTY.

andrewtoth commented on src/coins.cpp:
```diff
@@ -325,14 +321,14 @@ void CCoinsViewCache::SanityCheck() const
-        // Only 5 combinations are possible.
-        assert(attr != 2 && attr != 4 && attr != 7);
+        // Only 4 combinations are possible.
+        assert(attr != 2 && attr != 4 && attr != 6 && attr != 7);
```
Done.

andrewtoth commented on src/coins.cpp:
```diff
@@ -325,14 +322,14 @@ void CCoinsViewCache::SanityCheck() const
         if (entry.coin.IsSpent()) attr |= 4;
-        // Only 5 combinations are possible.
-        assert(attr != 2 && attr != 4 && attr != 7);
+        // Only 4 combinations are possible.
```
A coin cannot be spent and not DIRTY. I made a new commit to address this in AddCoin. The only time it will be in this state is when a `CCoinsViewCacheEntry` is constructed with an empty coin, which then has the coin moved into it a few lines later.

I think doing a follow up where we remove direct access to flags can remove a lot of code paths that we test for.

l0rinc replied:
That makes sense, as stated in other comments, I would be more comfortable doing that before this refactoring.

andrewtoth commented on src/coins.cpp:
```diff
@@ -49,10 +49,8 @@ CCoinsMap::iterator CCoinsViewCache::FetchCoin(const COutPoint &outpoint) const
-            ret->second.AddFlags(CCoinsCacheEntry::FRESH, *ret, m_sentinel);
-        }
+        // GetCoin will never return true if the coin is spent
+        Assume(!ret->second.coin.IsSpent());
```
It would be a fatal error when running in DEBUG mode, but is not compiled in for release.

> when do we throw and when do we Assume here?

I only added throws to `BatchWrite` because there are invalid states that can happen in our fuzz test harness which we want to handle properly. A follow up to remove setting the flags directly can help make that more clear and we wouldn't need to have the throws there anymore.

l0rinc replied:
Split out to https://github.com/bitcoin/bitcoin/pull/30906

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -892,8 +885,13 @@ BOOST_AUTO_TEST_CASE(ccoins_write)
             for (const char parent_flags : parent_value == ABSENT ? ABSENT_FLAGS : FLAGS)
-                for (const char child_flags : child_value == ABSENT ? ABSENT_FLAGS : CLEAN_FLAGS)
-                    CheckWriteCoins(parent_value, child_value, parent_value, parent_flags, child_flags, parent_flags);
+                for (const char child_flags : child_value == ABSENT ? ABSENT_FLAGS : CLEAN_FLAGS) {
```
I might be completely off here, but in case of e.g. `CheckWriteCoins(ABSENT, SPENT , FAIL  , NO_ENTRY   , DIRTY|FRESH, NO_ENTRY   );` we're calling `.IsDirty()` which returns false, but `m_next` and `m_prev` are set because `CheckWriteCoins` calls them with different flags.

Can this happen during actual usage, i.e. shouldn't they only be part of the internal linked list when they're dirty?

It bothers me a lot how many of the simulated combinations don't always seem to reflect real usage - while comments warn us of situations that may not be covered with tests at all.
Would it maybe make sense to fix the reliability of the tests before this PR?

l0rinc reviewed (changes requested):
The unreliability of many of our tests worries me, would you be open to fixing them in a PR before this one?

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -51,25 +51,19 @@ class CCoinsViewTest : public CCoinsView
             return false;
         }
         coin = it->second;
-        if (coin.IsSpent() && m_rng.randbool() == 0) {
```
Did this change require any other test adjustment?

andrewtoth replied:
No, but without this the next commit would break.

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -793,7 +787,7 @@ BOOST_AUTO_TEST_CASE(ccoins_add)
     CheckAddCoin(ABSENT, VALUE3, VALUE3, NO_ENTRY   , DIRTY|FRESH, false);
     CheckAddCoin(ABSENT, VALUE3, VALUE3, NO_ENTRY   , DIRTY      , true );
-    CheckAddCoin(SPENT , VALUE3, VALUE3, 0          , DIRTY|FRESH, false);
+    CheckAddCoin(SPENT , VALUE3, VALUE3, 0          , DIRTY      , false);
```
It seems to me that production code never actually calls `AddFlags` with `0` value - and we're always setting `DIRTY`, even when `FRESH` is set.
So it would make sense to change `AddFlags` from having a custom `uint8_t flags` parameter which is always `DIRTY` (and sometimes `FRESH`) in reality (and is only called with `0` flag in tests) to have a freshness parameter like `inline void SetDirtyAndFresh(CoinsCachePair& self, CoinsCachePair& sentinel, bool is_fresh = false) noexcept` instead.
This would enable us removing lots of invalid test cases like this one.

Like you've also mentioned, that would enable us removing the internal flags completely and instead of exposing the flags we could do the following instead:
```C++
inline bool IsDirty() const noexcept
{
    assert(!m_next == !m_prev);
    return !!m_next;
}
inline bool IsFresh() const noexcept { return m_is_fresh; }
```

andrewtoth replied:
Yes, absolutely! But, this PR or something else would need to be merged first, or we would need to have a method `SetFresh` because there's still a case where we have FRESH but not DIRTY.

I think we should keep the internal flags as a bitmap, since if we introduce new fields for each state it will increase the memory footprint of each `CoinsCachePair`.

l0rinc replied:
> because there's still a case where we have FRESH but not DIRTY.

I only see https://github.com/bitcoin/bitcoin/blob/34a101d4babd0cd2df86b414da385c3bded84027/src/coins.cpp#L204-L210 and https://github.com/bitcoin/bitcoin/blob/34a101d4babd0cd2df86b414da385c3bded84027/src/coins.cpp#L96 in production code - in both cases we're only setting `FRESH` together with `DIRTY`.
Did I miss anything in production code?

> I think we should keep the internal flags as a bitmap, since if we introduce new fields for each state it will increase the memory footprint of each CoinsCachePair.

I would rather we simplify to boolean, and if we need more flags, change it to an `uint8_t` again - I don't like over-generalizing for hypothetical scenarios - unless we're sure about that future direction (which I understood isn't the case).

andrewtoth replied:
This case will still have to be handled if this PR is not merged first:
https://github.com/bitcoin/bitcoin/blob/master/src/coins.cpp#L52-L55

> I would rather we simplify to boolean, and if we need more flags, change it to an uint8_t again

What I meant was adding a new field for DIRTY and a new field for FRESH. Since bools must be at least 1 byte it could make the pair larger depending on alignment.

l0rinc replied:
But we don't need `bool` for dirty anymore, only for freshness - so we can even get rid of the enum (moving it over to the test maybe). I'll experiment to see what the next smallest change is before this PR, we need to have tests that are more representative of what's actually possible.

l0rinc commented on src/coins.cpp:
```diff
@@ -341,7 +337,7 @@ void CCoinsViewCache::SanityCheck() const
-        assert(it->second.IsDirty() || it->second.IsFresh());
+        assert(it->second.IsDirty());
         // Count the number of entries actually in the list.
         ++count_linked;
```
to avoid an infinite loop when the linking is incorrect (e.g. next returns itself), we could `assert(count_linked++ < count_flagged)`;

andrewtoth replied:
This is unrelated to this change. It will timeout if that happens, but I suppose a separate PR could be made to add that.

l0rinc reviewed (commented):
I went over the changes again after the rebase. I like that the change is made in small commits, but they're not telling me a story yet, I don't fully understand why some changes were split out and others done in the same commit.
These refactorings are really important to make coins logic accessible again to all.
Many of my previous questions still apply and I would need to understand them before I can ack.

andrewtoth commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -151,7 +151,9 @@ FUZZ_TARGET(coins_view, .init = initialize_coins_view)
                     expected_code_path = true;
                 } catch (const std::logic_error& e) {
-                    if (e.what() == std::string{"FRESH flag misapplied to coin that exists in parent cache"}) {
+                    if (e.what() == std::string{"FRESH flag misapplied to coin that exists in parent cache"}
```
I think after this change is merged, we can remove `SetFresh` and then the fuzz harness will not be able to add a FRESH-but-not-DIRTY. Then we can remove these catches and actually break if the fuzzer reaches them.

Right now the fuzz harness can still call `SetFresh` without calling `SetDirty`.

andrewtoth replied:
The other 2 though I think you're right. They should not be possible to hit even in the fuzz harness now. Will remove and fuzz.

andrewtoth replied:
Looking again, these can still be called with the fuzz harness. Removing `SetFresh` in a follow-up would allow us to remove one of them, but the other 2 would not since the `random_coin` can be a spent coin.

l0rinc replied:
> since the random_coin can be a spent coin

But can that happen in reality or are we testing something that's difficult to justify?

andrewtoth replied:
I removed `SetFresh` in the last commit. This lets us remove these tests.

andrewtoth commented on src/test/coins_tests.cpp:
```diff
@@ -58,14 +58,12 @@ class CCoinsViewTest : public CCoinsView
+            assert(it->second.IsDirty());
+            map_[it->first] = it->second.coin;
+            if (it->second.coin.IsSpent() && m_rng.randrange(3) == 0) {
+                // Randomly delete empty entries on write.
```
Spent entries are empty. I guess we could unify the terminology here.

andrewtoth replied:
Done.

andrewtoth commented on src/txdb.cpp:
```diff
@@ -147,7 +145,7 @@ bool CCoinsViewDB::BatchWrite(CoinsViewCacheCursor& cursor, const uint256 &hashB
     LogDebug(BCLog::COINDB, "Writing final batch of %.2f MiB\n", batch.SizeEstimate() * (1.0 / 1048576.0));
     bool ret = m_db->WriteBatch(batch);
-    LogDebug(BCLog::COINDB, "Committed %u changed transaction outputs (out of %u) to coin database...\n", (unsigned int)changed, (unsigned int)count);
+    LogDebug(BCLog::COINDB, "Committed %u changed transaction outputs to coin database...\n", (unsigned int)count);
```
I don't think we can get rid of newlines...

l0rinc replied:
It was removed in https://github.com/bitcoin/bitcoin/pull/30929/files#diff-8c9b223f6f7e311ef2a0585813c17e4cc78119d39ef9956431164336e9158832R137 - setting it back demonstrates that it's not needed anymore (but tolerated):
```patch
diff --git a/src/test/logging_tests.cpp b/src/test/logging_tests.cpp
--- a/src/test/logging_tests.cpp	(revision ac69069c050a3a3a34e2b9f27a458872c4704e42)
+++ b/src/test/logging_tests.cpp	(date 1733497035556)
@@ -134,7 +134,7 @@
 BOOST_FIXTURE_TEST_CASE(logging_LogPrintMacros, LogSetup)
 {
     LogTrace(BCLog::NET, "foo6: %s", "bar6"); // not logged
-    LogDebug(BCLog::NET, "foo7: %s", "bar7");
+    LogDebug(BCLog::NET, "foo7: %s\n", "bar7");
     LogInfo("foo8: %s", "bar8");
     LogWarning("foo9: %s", "bar9");
     LogError("foo10: %s", "bar10");

```

andrewtoth replied:
Done.

andrewtoth commented on src/coins.cpp:
```diff
@@ -278,7 +275,7 @@ bool CCoinsViewCache::Sync()
 void CCoinsViewCache::Uncache(const COutPoint& hash)
 {
     CCoinsMap::iterator it = cacheCoins.find(hash);
-    if (it != cacheCoins.end() && !it->second.IsDirty() && !it->second.IsFresh()) {
+    if (it != cacheCoins.end() && !it->second.IsDirty()) {
```
Just seems like noise. What is the benefit of this refactor?

andrewtoth replied:
Done.

andrewtoth commented on src/coins.cpp:
```diff
@@ -341,7 +338,7 @@ void CCoinsViewCache::SanityCheck() const
         assert(it->second.Prev()->second.Next() == it);
         // Verify they are actually flagged.
-        assert(it->second.IsDirty() || it->second.IsFresh());
+        assert(it->second.IsDirty());
```
We only `Assume` in prev and next, here we assert.

andrewtoth commented on src/coins.cpp:
```diff
@@ -79,20 +79,18 @@ void CCoinsViewCache::AddCoin(const COutPoint &outpoint, Coin&& coin, bool possi
+        // A spent FRESH coin cannot exist in the cache because a FRESH coin
+        // is simply erased when it is spent.
         //
         // Re-adding a spent coin can happen in the case of a re-org (the coin
```
Yes, it can happen during a reorg. But we know we can't have a spent coin that is not DIRTY, so the comments being updated here are making this not be redundant. When the coin is spent in `DisconnectBlock`, it is either erased if FRESH or set DIRTY. The only time we would get into this block if the coin is not DIRTY is if it was just inserted.

andrewtoth commented on src/coins.cpp:
```diff
@@ -81,20 +79,18 @@ void CCoinsViewCache::AddCoin(const COutPoint &outpoint, Coin&& coin, bool possi
-        // DIRTY, then it can be marked FRESH.
-        fresh = !it->second.IsDirty();
+        // If the coin doesn't exist in the current cache then it can be marked FRESH.
+        fresh = inserted;
```
Let's leave this for a follow-up.

andrewtoth commented on src/test/coins_tests.cpp:
```diff
@@ -793,7 +787,7 @@ BOOST_AUTO_TEST_CASE(ccoins_add)
      */
```
I'm not sure. It might just be redundant to our current fuzz harness. It's also valuable to have these tests run on every CI.

andrewtoth commented on src/test/fuzz/coinscache_sim.cpp:
```diff
@@ -148,8 +146,7 @@ class CoinsViewBottom final : public CCoinsView
+        if (auto it = m_data.find(outpoint); it != m_data.end() && !it->second.IsSpent()) return it->second;
         return std::nullopt;
     }
 
```
Done.

andrewtoth commented on src/test/fuzz/coinscache_sim.cpp:
```diff
@@ -148,8 +146,7 @@ class CoinsViewBottom final : public CCoinsView
 public:
     std::optional<Coin> GetCoin(const COutPoint& outpoint) const final
     {
-        // TODO GetCoin shouldn't return spent coins
-        if (auto it = m_data.find(outpoint); it != m_data.end()) return it->second;
+        if (auto it = m_data.find(outpoint); it != m_data.end() && !it->second.IsSpent()) return it->second;
         return std::nullopt;
     }
```
We would then have to share code between fuzz and unit tests, right? I'm not sure it's worth it for this.

l0rinc replied:
I meant at least stylistically (helps with finding problems if the same logic is written in the same way)

andrewtoth commented on src/coins.cpp:
```diff
@@ -49,12 +49,9 @@ CCoinsMap::iterator CCoinsViewCache::FetchCoin(const COutPoint &outpoint) const
     const auto [ret, inserted] = cacheCoins.try_emplace(outpoint);
     if (inserted) {
         if (auto coin{base->GetCoin(outpoint)}) {
+            Assume(!coin.value().IsSpent());
```
Removed that line.

andrewtoth commented on src/coins.h:
```diff
@@ -188,14 +181,14 @@ struct CCoinsCacheEntry
+    //! Only call Next when this entry is DIRTY
     CoinsCachePair* Next() const noexcept
     {
         Assume(m_flags);
```
Done.

l0rinc commented on src/test/fuzz/coinscache_sim.cpp:
```diff
@@ -148,14 +146,13 @@ class CoinsViewBottom final : public CCoinsView
     {
-        // TODO GetCoin shouldn't return spent coins
-        if (auto it = m_data.find(outpoint); it != m_data.end()) return it->second;
+        if (auto it = m_data.find(outpoint); it != m_data.end() && !it->second.IsSpent()) return it->second;
```
nit: style of this `GetCoin` is still a tiny bit different from the other one (i.e. brace init)

andrewtoth replied:
Done.

l0rinc commented on src/coins.cpp:
```diff
@@ -315,14 +315,14 @@ void CCoinsViewCache::SanityCheck() const
 
         // Count the number of entries we expect in the linked list.
-        if (entry.IsDirty() || entry.IsFresh()) ++count_flagged;
+        if (entry.IsDirty()) ++count_flagged;
```
same: this [commit](https://github.com/bitcoin/bitcoin/pull/30673/commits/2c3e62d4526809e9d2f5a01ff4d049fb6754cce6) seems related to the previous commit, can we merge them?

l0rinc commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -151,7 +151,9 @@ FUZZ_TARGET(coins_view, .init = initialize_coins_view)
+                        || e.what() == std::string{"A FRESH coin was not removed when it was spent"}
+                        || e.what() == std::string{"A non-DIRTY coin was returned from the cursor in BatchWrite"}) {
                         expected_code_path = true;
                     }
```
We may want to add a `Flush` here as well: https://github.com/bitcoin/bitcoin/pull/28216/files#diff-1ef3b6a1936b50f3d5ec4a1786d9e2d63d1a3e1815b103e67f20601995f355b4R79

andrewtoth commented on src/test/coins_tests.cpp:
```diff
@@ -48,10 +48,8 @@ class CCoinsViewTest : public CCoinsView
-            }
+        if (auto it{map_.find(outpoint)}; it != map_.end() && !it->second.IsSpent()) {
+            return it->second;
         }
```
Done.

andrewtoth commented on src/coins.h:
```diff
@@ -102,7 +102,6 @@ using CoinsCachePair = std::pair<const COutPoint, CCoinsCacheEntry>;
  * - unspent, FRESH, DIRTY (e.g. a new coin created in the cache)
  * - unspent, not FRESH, DIRTY (e.g. a coin changed in the cache during a reorg)
  * - unspent, not FRESH, not DIRTY (e.g. an unspent coin fetched from the parent cache)
- * - spent, FRESH, not DIRTY (e.g. a spent coin fetched from the parent cache)
```
Done.

andrewtoth commented on src/coins.h:
```diff
@@ -195,7 +188,7 @@ struct CCoinsCacheEntry
     }
 
-    //! Only call Prev when this entry is DIRTY, FRESH, or both
+    //! Only call Next when this entry is DIRTY, FRESH, or both
```
Fixed.

andrewtoth commented on src/coins.cpp:
```diff
@@ -311,12 +311,13 @@ void CCoinsViewCache::SanityCheck() const
     size_t recomputed_usage = 0;
     size_t count_flagged = 0;
     for (const auto& [_, entry] : cacheCoins) {
```
I merged it with https://github.com/bitcoin/bitcoin/pull/30673/commits/aaef2d6c0d801dc48798e9a7d0911c234870412e instead.

andrewtoth commented:
@l0rinc rebased. Thank you for your reviews! I have tried to address all your comments. There are some older comments that I'm unsure are still relevant. Please let me know if there is still anything outstanding that needs to be addressed.

andrewtoth committed:
coins: remove SetFresh and replace with boolean flag passed to SetDirty

It is not possible to have a FRESH CCoinsCacheEntry that is not also
DIRTY. Test code uses the SetFresh method out of order to simulate
entries that are FRESH but not DIRTY. By removing the method entirely
we can simplify the test code and make the production code easier to
understand.

l0rinc commented on src/coins.h:
```diff
@@ -174,8 +167,10 @@ struct CCoinsCacheEntry
 
-    static void SetDirty(CoinsCachePair& pair, CoinsCachePair& sentinel) noexcept { AddFlags(DIRTY, pair, sentinel); }
-    static void SetFresh(CoinsCachePair& pair, CoinsCachePair& sentinel) noexcept { AddFlags(FRESH, pair, sentinel); }
+    static void SetDirty(CoinsCachePair& pair, CoinsCachePair& sentinel, bool fresh = false) noexcept
+    {
+        AddFlags(fresh ? FRESH | DIRTY : DIRTY, pair, sentinel);
+    }
```
Since it's only called here, we could inline and simplify it now:
```suggestion
    static void SetDirty(CoinsCachePair& pair, CoinsCachePair& sentinel, bool fresh = false) noexcept
    {
        if (!pair.second.m_flags) {
            Assume(!pair.second.m_prev && !pair.second.m_next);
            pair.second.m_prev = sentinel.second.m_prev;
            pair.second.m_next = &sentinel;
            sentinel.second.m_prev = &pair;
            pair.second.m_prev->second.m_next = &pair;
        }
        Assume(pair.second.m_prev && pair.second.m_next);
        pair.second.m_flags |= fresh ? FRESH | DIRTY : DIRTY;
    }
```

----

Not sure, but do the flags still make sense after this or would a `bool is_fresh` also suffice? The dirtiness can be deduced from the `m_next`/`m_prev` pair so the only unknown is the freshness which doesn't necessitate a flag anymore - right?

l0rinc commented on src/coins.h:
```diff
@@ -188,17 +183,15 @@ struct CCoinsCacheEntry
     bool IsDirty() const noexcept { return m_flags & DIRTY; }
     bool IsFresh() const noexcept { return m_flags & FRESH; }
```
we don't have Fresh anymore, only `IsDirtyAndFresh`, right?
```suggestion
    bool IsDirtyAndFresh() const noexcept
    {
        const bool is_fresh = m_flags & FRESH;
        Assume(IsDirty() || !is_fresh);
        return is_fresh;
    }
```

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -48,26 +48,20 @@ class CCoinsViewTest : public CCoinsView
-                return it->second; // TODO spent coins shouldn't be returned
-            }
-        }
+        if (auto it{map_.find(outpoint)}; it != map_.end() && !it->second.IsSpent()) return it->second;
```
Given the lack of reviews, we may want to split out the test cleanups (removing invalid states) and refactors to a separate PR, where this will be a tracking PR for where we want to get to in safer/smaller steps

l0rinc commented on src/test/coinscachepair_tests.cpp:
```diff
@@ -154,22 +154,14 @@ BOOST_AUTO_TEST_CASE(linked_list_set_state)
-    CCoinsCacheEntry::SetFresh(n2, sentinel);
-    BOOST_CHECK(n2.second.IsFresh() && !n2.second.IsDirty());
+    // Check that setting DIRTY and FRESH on new node inserts it after n1
+    CCoinsCacheEntry::SetDirty(n2, sentinel, true);
```
```suggestion
    CCoinsCacheEntry::SetDirty(n2, sentinel, /*fresh*/true);
```

l0rinc commented on src/test/fuzz/coinscache_sim.cpp:
```diff
@@ -148,14 +146,13 @@ class CoinsViewBottom final : public CCoinsView
     bool HaveCoin(const COutPoint& outpoint) const final
     {
-        return m_data.count(outpoint);
+        return GetCoin(outpoint).has_value();
```
we should be able to do this safely in a prefactor PR regardless of the other changes

l0rinc reviewed (commented):
The `src/coins.cpp` prod changes are a bit scary - while it's a very important change, based on the lack of enthusiasm by others, we may want to separate the tests and small refactors into PRs that will get us closer to this.

One could remove the invalid test states - without removing the invalid production code states, just potentially adding TODOs there in case it's missing.
The `SetFresh` and flag removal could be next and if all of those are merged, we can continue with the scary `FetchCoin`, `AddCoin`, `BatchWrite` and `Uncache`.

Left a few new comments quickly, let me know what you think.

andrewtoth commented:
Closing in favor of https://github.com/bitcoin/bitcoin/pull/33018.
